pytorch_lightning/trainer/trainer.py

```diff
@@ -500,20 +500,23 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
         """
         try:
             return trainer_fn(*args, **kwargs)
-        except KeyboardInterrupt:
+        # TODO(@daniellepintz): treat KeyboardInterrupt as BaseException (delete the code below) in v1.7
+        except KeyboardInterrupt as exception:
             rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
             # user could press Ctrl+c many times... only shutdown once
             if not self.interrupted:
                 self.state.status = TrainerStatus.INTERRUPTED
                 self.on_keyboard_interrupt()
-        except BaseException:
+                self.on_exception(exception)
+        except BaseException as exception:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
                 # try syncing remaing processes, kill otherwise
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             self._on_exception()
             # reset bookkeeping
             self.state.stage = None
+            self.on_exception(exception)
```

Comment:
it's not something we can solve in this PR, but it's very confusing that there's `self._on_exception` and `self.on_exception` which do completely different things. 

my preference is that we remove the trainer callback hook mixin so that callback hooks aren't automatically re-exposed on the trainer this way. https://github.com/PyTorchLightning/pytorch-lightning/blob/46b00a7de44d226a3da22b258c74cc2b3cd92bec/pytorch_lightning/trainer/callback_hook.py#L27

but this is something we'll need to follow up on separately

Reply:
yeah I completely agree about _on_exception. Not sure if I'm following you - do you mean remove TrainerCallbackHookMixin and bring all the callback hooks to trainer.py, and then merge _on_exception and on_exception in the trainer?